File: backend/java-spring/src/main/java/com/unosquare/worldcup/util/CostCalculator.java
package com.unosquare.worldcup.util;

import com.unosquare.worldcup.dto.*;
import com.unosquare.worldcup.model.City;
import com.unosquare.worldcup.model.FlightPrice;
import com.unosquare.worldcup.util.BuildRouteUtil;
import org.springframework.stereotype.Component;

import java.time.temporal.ChronoUnit;
import java.util.*;
import java.util.stream.Collectors;

/**
 * CostCalculator — YOUR TASK #5
 *
 * Calculates trip costs for a given set of matches.
 *
 * ============================================================
 * WHAT YOU NEED TO IMPLEMENT:
 * ============================================================
 *
 * The calculate() method — Build the budget result by:
 *   1. Finding which countries are visited and which are missing
 *   2. Calculating total costs using the helper methods
 *   3. Building the cost breakdown DTO
 *   4. Determining if the trip is feasible
 *   5. Generating suggestions if not feasible
 *   6. Building and returning the final result
 *
 * ============================================================
 * HELPER METHODS PROVIDED (no changes needed):
 * ============================================================
 *
 * - calculateTicketsCost(matches) → total ticket cost
 * - calculateFlightsCost(originCity, matches, flightPrices) → total flight cost
 * - calculateAccommodationCost(matches) → total accommodation cost
 * - getMatchTicketPrice(match) → price of a single match ticket
 * - BuildRouteUtil.buildRoute(matches, strategy) → builds the route DTO
 *
 * ============================================================
 * CONSTRAINT:
 * ============================================================
 *
 * The user must attend at least 1 match in each country (USA, Mexico, Canada).
 * If any country is missing, the trip is NOT feasible.
 *
 */
@Component
public class CostCalculator {

    private static final Set<String> REQUIRED_COUNTRIES = Set.of("USA", "Mexico", "Canada");

    // ============================================================
    //  Calculate budget result
    // ============================================================
    //
    // TODO: Implement this method
    //
    // Parameters:
    //   matches       → List of matches to include in the trip
    //   budget        → The user's budget
    //   originCityId  → ID of the starting city
    //   flightPrices  → List of flight prices between cities
    //   originCity    → The starting city object
    //
    // Steps:
    //   1. Sort matches by kickoff date
    //   2. Find countries visited (from match cities)
    //   3. Find missing countries (compare against REQUIRED_COUNTRIES)
    //   4. Calculate costs using helper methods:
    //      - ticketsCost = calculateTicketsCost(sortedMatches)
    //      - flightsCost = calculateFlightsCost(originCity, sortedMatches, flightPrices)
    //      - accommodationCost = calculateAccommodationCost(sortedMatches)
    //   5. Build CostBreakdownDTO with the costs
    //   6. Determine feasibility: no missing countries AND totalCost <= budget
    //   7. Generate suggestions using generateSuggestions() helper
    //   8. Build route using BuildRouteUtil.buildRoute(sortedMatches, "budget-optimised")
    //   9. Return BudgetResultDTO with all the data
    //
    // ============================================================
    public BudgetResultDTO calculate(
            List<MatchWithCityDTO> matches,
            Double budget,
            String originCityId,
            List<FlightPrice> flightPrices,
            City originCity
    ) {
        // TODO: Implement this method
        //
        // Steps:
        //   1. Sort matches by kickoff date
        //   2. Find countries visited (from match cities)
        //   3. Find missing countries (compare against REQUIRED_COUNTRIES)
        //   4. Calculate costs using helper methods:
        //      - ticketsCost = calculateTicketsCost(sortedMatches)
        //      - flightsCost = calculateFlightsCost(originCity, sortedMatches, flightPrices)
        //      - accommodationCost = calculateAccommodationCost(sortedMatches)
        //   5. Build CostBreakdownDTO with the costs
        //   6. Determine feasibility: no missing countries AND totalCost <= budget
        //   7. Generate suggestions using generateSuggestions() helper
        //   8. Build route using BuildRouteUtil.buildRoute(sortedMatches, "budget-optimised")
        //   9. Return BudgetResultDTO with all the data
        //
        return null;
    }

    /**
     * Generate helpful suggestions when the trip is not feasible.
     * (Provided — no changes needed)
     */
    private List<String> generateSuggestions(
            List<String> missingCountries,
            double totalCost,
            double budget,
            List<MatchWithCityDTO> matches
    ) {
        List<String> suggestions = new ArrayList<>();

        if (!missingCountries.isEmpty()) {
            suggestions.add("Add matches in: " + String.join(", ", missingCountries));
        }

        if (totalCost > budget) {
            suggestions.add(String.format("You need $%.2f more to complete this trip", totalCost - budget));

            // Find most expensive match
            Optional<MatchWithCityDTO> mostExpensive = matches.stream()
                    .max(Comparator.comparingDouble(m -> getMatchTicketPrice(m)));
            mostExpensive.ifPresent(m -> {
                double price = getMatchTicketPrice(m);
                suggestions.add(String.format("Consider removing %s vs %s to save $%.2f on tickets",
                        m.getHomeTeam().getName(), m.getAwayTeam().getName(), price));
            });
        }

        return suggestions;
    }

    private double calculateTicketsCost(List<MatchWithCityDTO> matches) {
        return matches.stream()
                .mapToDouble(this::getMatchTicketPrice)
                .sum();
    }

    private double getMatchTicketPrice(MatchWithCityDTO match) {
        Double ticketPrice = match.getTicketPrice();
        return ticketPrice != null ? ticketPrice : 150.0; // Default if not set
    }

    private double calculateFlightsCost(City originCity, List<MatchWithCityDTO> matches, List<FlightPrice> flightPrices) {
        if (matches.isEmpty()) return 0;

        double totalFlightCost = 0;

        // Flight from origin to first match
        City firstMatchCity = matches.get(0).getCity();
        totalFlightCost += getFlightPrice(originCity, firstMatchCity, flightPrices);

        // Flights between consecutive matches
        for (int i = 0; i < matches.size() - 1; i++) {
            City fromCity = matches.get(i).getCity();
            City toCity = matches.get(i + 1).getCity();
            if (!fromCity.getId().equals(toCity.getId())) {
                totalFlightCost += getFlightPrice(fromCity, toCity, flightPrices);
            }
        }

        return totalFlightCost;
    }

    private double getFlightPrice(City from, City to, List<FlightPrice> flightPrices) {
        if (from.getId().equals(to.getId())) return 0;

        // Look for exact price in flight prices
        Optional<FlightPrice> exactPrice = flightPrices.stream()
                .filter(fp -> fp.getOriginCity().getId().equals(from.getId())
                        && fp.getDestinationCity().getId().equals(to.getId()))
                .findFirst();

        if (exactPrice.isPresent()) {
            return exactPrice.get().getPriceUsd();
        }

        // Estimate based on distance if not found (roughly $0.10 per km)
        double distance = calculateDistance(
                from.getLatitude(), from.getLongitude(),
                to.getLatitude(), to.getLongitude()
        );
        return Math.round(distance * 0.10 * 100.0) / 100.0;
    }

    private double calculateDistance(double lat1, double lon1, double lat2, double lon2) {
        final double EARTH_RADIUS_KM = 6371.0;
        double dLat = Math.toRadians(lat2 - lat1);
        double dLon = Math.toRadians(lon2 - lon1);
        double a = Math.sin(dLat / 2) * Math.sin(dLat / 2)
                + Math.cos(Math.toRadians(lat1))
                * Math.cos(Math.toRadians(lat2))
                * Math.sin(dLon / 2) * Math.sin(dLon / 2);
        double c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
        return EARTH_RADIUS_KM * c;
    }

    private double calculateAccommodationCost(List<MatchWithCityDTO> matches) {
        if (matches.size() < 2) {
            // At least one night for a single match
            if (matches.size() == 1) {
                return matches.get(0).getCity().getAccommodationPerNight();
            }
            return 0;
        }

        double totalAccommodation = 0;

        for (int i = 0; i < matches.size() - 1; i++) {
            MatchWithCityDTO currentMatch = matches.get(i);
            MatchWithCityDTO nextMatch = matches.get(i + 1);

            long nights = ChronoUnit.DAYS.between(
                    currentMatch.getKickoff().toLocalDate(),
                    nextMatch.getKickoff().toLocalDate()
            );

            // Ensure at least 1 night
            nights = Math.max(1, nights);

            double nightlyRate = currentMatch.getCity().getAccommodationPerNight();
            totalAccommodation += nights * nightlyRate;
        }

        // Add one night for the last city
        totalAccommodation += matches.get(matches.size() - 1).getCity().getAccommodationPerNight();

        return totalAccommodation;
    }
}

File: backend/node-express/src/utils/CostCalculator.ts
import {
  MatchWithCity,
  FlightPrice,
  BudgetResult,
  CostBreakdown,
  City,
} from '../strategies/RouteStrategy';
import { buildRoute } from './buildRoute';
import { calculateDistance } from './haversine';

/**
 * CostCalculator — YOUR TASK #5
 *
 * Calculates trip costs for a given set of matches.
 *
 * ============================================================
 * WHAT YOU NEED TO IMPLEMENT:
 * ============================================================
 *
 * The calculate() method — Build the budget result by:
 *   1. Finding which countries are visited and which are missing
 *   2. Calculating total costs using the helper methods
 *   3. Building the cost breakdown
 *   4. Determining if the trip is feasible
 *   5. Generating suggestions if not feasible
 *   6. Building and returning the final result
 *
 * ============================================================
 * HELPER METHODS PROVIDED (no changes needed):
 * ============================================================
 *
 * - calculateTicketsCost(matches) → total ticket cost
 * - calculateFlightsCost(originCity, matches, flightPrices) → total flight cost
 * - calculateAccommodationCost(matches) → total accommodation cost
 * - getFlightPrice(from, to, flightPrices) → price of a single flight
 * - generateSuggestions(...) → helpful suggestions when over budget
 * - buildRoute(matches, strategy) → builds the route DTO
 *
 * ============================================================
 * CONSTRAINT:
 * ============================================================
 *
 * The user must attend at least 1 match in each country (USA, Mexico, Canada).
 * If any country is missing, the trip is NOT feasible.
 *
 */

const REQUIRED_COUNTRIES = ['USA', 'Mexico', 'Canada'];

// ============================================================
//  Calculate budget result
// ============================================================
//
// TODO: Implement this function
//
// Parameters:
//   matches       → List of matches to include in the trip
//   budget        → The user's budget
//   originCityId  → ID of the starting city
//   flightPrices  → List of flight prices between cities
//   originCity    → The starting city object
//
// Steps:
//   1. Sort matches by kickoff date
//   2. Find countries visited (from match cities)
//   3. Find missing countries (compare against REQUIRED_COUNTRIES)
//   4. Calculate costs using helper methods:
//      - ticketsCost = calculateTicketsCost(sortedMatches)
//      - flightsCost = calculateFlightsCost(originCity, sortedMatches, flightPrices)
//      - accommodationCost = calculateAccommodationCost(sortedMatches)
//   5. Build CostBreakdown with the costs
//   6. Determine feasibility: no missing countries AND totalCost <= budget
//   7. Generate suggestions using generateSuggestions() helper
//   8. Build route using buildRoute(sortedMatches, 'budget-optimised')
//   9. Return BudgetResult with all the data
//
// ============================================================

export function calculate(
  matches: MatchWithCity[],
  budget: number,
  originCityId: string,
  flightPrices: FlightPrice[],
  originCity: City
): BudgetResult {

  // Edge case handling: If no matches are provided, exit early.
  if (!matches || matches.length === 0) {
    return {
      feasible: false,
      costBreakdown: { flights: 0, accommodation: 0, tickets: 0, total: 0 },
      countriesVisited: [],
      missingCountries: [...REQUIRED_COUNTRIES],
      suggestions: ['No matches selected. Please select at least one match in each required country.'],
    };
  }

  // 1. Sort matches chronologically by kickoff date
  // Again the spread operator can be used to create a shallow copy, ensuring we dont mutate the original array
  const sortedMatches = [...matches].sort(
    (a, b) => new Date(a.kickoff).getTime() - new Date(b.kickoff).getTime()
  );

  // 2. Find countries visited
  // A set can be used to automatically handle deduplication, i.e. multiple visits to the USA only stores the country once
  const visitedSet = new Set<string>();
  for (const match of sortedMatches) {
    visitedSet.add(match.city.country);
  }
  const countriesVisited = Array.from(visitedSet);

  // 3.Find missing countries
  const missingCountries = REQUIRED_COUNTRIES.filter(country => !visitedSet.has(country));

  // 4. Calculate individual costs using the provided helper methods
  const ticketsCost = calculateTicketsCost(sortedMatches);
  const flightsCost = calculateFlightsCost(originCity, sortedMatches, flightPrices);
  const accommodationCost = calculateAccommodationCost(sortedMatches);

  const totalCost = ticketsCost + flightsCost + accommodationCost;

  // 5. Build CostBreakdown object
  const costBreakdown: CostBreakdown = {
    flights: flightsCost,
    accommodation: accommodationCost,
    tickets: ticketsCost,
    total: totalCost,
  };

  // 6. Determine feasibility: There must be no countries missing && total cost must fit within the budget
  const feasible = (missingCountries.length === 0) && (totalCost <= budget);

  // 7. Generate helpful suggestions (especially if they are over budget) 
  const suggestions = generateSuggestions(missingCountries, totalCost, budget, sortedMatches);

  // 8. Build the route object representing the physical travel path
  const route = buildRoute(sortedMatches, 'budget-optimised');

  // 9 Return the final budgetResult
  return {
    feasible,
    route,
    costBreakdown,
    countriesVisited,
    missingCountries,
    // Provide the minimum required budget only if they exceeded it
    minimumBudgetRequired: totalCost > budget ? totalCost : undefined,
    suggestions,
  };
}

// ============================================================
//  Helper Methods (provided - no changes needed)
// ============================================================

/**
 * Generate helpful suggestions when the trip is not feasible.
 */
function generateSuggestions(
  missingCountries: string[],
  totalCost: number,
  budget: number,
  matches: MatchWithCity[]
): string[] {
  const suggestions: string[] = [];

  if (missingCountries.length > 0) {
    suggestions.push(`Add matches in: ${missingCountries.join(', ')}`);
  }

  if (totalCost > budget) {
    suggestions.push(`You need $${(totalCost - budget).toFixed(2)} more to complete this trip`);

    // Find most expensive match
    const sortedByPrice = [...matches].sort(
      (a, b) => getMatchTicketPrice(b) - getMatchTicketPrice(a)
    );
    if (sortedByPrice.length > 0) {
      const mostExpensive = sortedByPrice[0];
      const price = getMatchTicketPrice(mostExpensive);
      suggestions.push(
        `Consider removing ${mostExpensive.homeTeam.name} vs ${mostExpensive.awayTeam.name} to save $${price.toFixed(2)} on tickets`
      );
    }
  }

  return suggestions;
}

function calculateTicketsCost(matches: MatchWithCity[]): number {
  return matches.reduce((sum, match) => sum + getMatchTicketPrice(match), 0);
}

function getMatchTicketPrice(match: MatchWithCity): number {
  return match.ticketPrice ?? 150.0; // Default if not set
}

function calculateFlightsCost(
  originCity: City,
  matches: MatchWithCity[],
  flightPrices: FlightPrice[]
): number {
  if (matches.length === 0) return 0;

  let totalFlightCost = 0;

  // Flight from origin to first match
  const firstMatchCity = matches[0].city;
  totalFlightCost += getFlightPrice(originCity, firstMatchCity, flightPrices);

  // Flights between consecutive matches
  for (let i = 0; i < matches.length - 1; i++) {
    const fromCity = matches[i].city;
    const toCity = matches[i + 1].city;
    if (fromCity.id !== toCity.id) {
      totalFlightCost += getFlightPrice(fromCity, toCity, flightPrices);
    }
  }

  return totalFlightCost;
}

function getFlightPrice(from: City, to: City, flightPrices: FlightPrice[]): number {
  if (from.id === to.id) return 0;

  // Look for exact price in flight prices
  const exactPrice = flightPrices.find(
    (fp) => fp.origin_city_id === from.id && fp.destination_city_id === to.id
  );

  if (exactPrice) {
    return exactPrice.price_usd;
  }

  // Estimate based on distance if not found (roughly $0.10 per km)
  const distance = calculateDistance(from.latitude, from.longitude, to.latitude, to.longitude);
  return Math.round(distance * 0.1 * 100) / 100;
}

function calculateAccommodationCost(matches: MatchWithCity[]): number {
  if (matches.length < 2) {
    // At least one night for a single match
    if (matches.length === 1) {
      return matches[0].city.accommodation_per_night;
    }
    return 0;
  }

  let totalAccommodation = 0;

  for (let i = 0; i < matches.length - 1; i++) {
    const currentMatch = matches[i];
    const nextMatch = matches[i + 1];

    const currentDate = new Date(currentMatch.kickoff);
    const nextDate = new Date(nextMatch.kickoff);
    let nights = Math.ceil(
      (nextDate.getTime() - currentDate.getTime()) / (1000 * 60 * 60 * 24)
    );

    // Ensure at least 1 night
    nights = Math.max(1, nights);

    const nightlyRate = currentMatch.city.accommodation_per_night;
    totalAccommodation += nights * nightlyRate;
  }

  // Add one night for the last city
  totalAccommodation += matches[matches.length - 1].city.accommodation_per_night;

  return totalAccommodation;
}
